lib/puppet/datatypes.rb

```diff
@@ -0,0 +1,190 @@
+# Data types in the Puppet Language can have implementations written in Ruby
+# and distributed in puppet modules. A data type can be declared together with
+# its implementation by creating a file in 'lib/puppet/functions/<modulename>'.
+# The name of the file must be the downcased name of the data type followed by
+# the extension '.rb'.
+#
+# A data type is created by calling {Puppet::DataTypes.create_type(<type name>)}
+# and passing it a block that defines the data type interface and implementation.
+#
+# Data types are namespaced inside the modules that contains them. The name of the
+# data type is prefixed with the name of the module. As with all type names, each
+# segment of the name must start with an uppercase letter.
+#
+# @example A simple data type
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String
+#       }
+#     PUPPET
+#   end
+#
+# The above example does not declare an implementation which makes it exactly similar
+# to adding the following contents in a file named 'user.pp' under the 'types' directory
+# of the module root.
+#
+#   type Auth::User = Object[
+#     attributes => {
+#       name => String,
+#       email => String
+#     }]
+#
+# Both declarations are valid and will be found by the module loader.
+#
+# Anatomy of a data type
+# ---
+#
+# Data types consist of an interface and an implementation. A data type can often be used
+# by only declaring the interface. Unless an implementation is found it will be automatically
+# generated.
+#
+# In some cases, the generated implementation is not enough. In other cases it might be
+# necessary to declare a data type that maps to an already existing implementation.
+#
+# @example Adding implementation on top of the generated type using `implementation`
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String,
+#         year_of_birth => Integer,
+#         age => { type => Integer, kind => derived }
+#       }
+#       PUPPET
+#
+#     implementation do
+#       def age
+#         DateTime.now.year - @year_of_birth
+#       end
+#     end
+#   end
+#
+# @example Appointing an already existing implementation class
+#
+# Assumes the following class is declared under 'lib/auth/utils' in the module:
```

Comment:
IIRC we recommend that 3d party ruby code / helpers like this should be placed under the puppet-x name space, and then with author unless it is already in a well established name space (i.e. in Ruby or some other gem). (I think the name is on the form  `PuppetX` - but don't remember if it  is some other variation).